### PR TITLE
add delegate to track events from the dependency resolver

### PR DESCRIPTION
**Pull request by tomerd**
motivation:
* significant part of package resolution can be spent in the dependency resolver which does not emit events we can use to tell the user what is happening
* a follow up PR would use these new events to emit information when package dependencies are resolved, e.g when running swift package update

chnages:
* define DependencyResolverDelegate
* replace logging in PubgrubDependencyResolver with calling the relevant delegate methods
* add TracingDependencyResolverDelegate which back-fills the existing logging behavior in PubgrubDependencyResolver
* adjust call-sites

### CODE REVIEW

**Comment by tomerd** [comment]
@neonichu @abertelrud looking for early feedback on the approach and open questions, so we can finalize the feature

File: Sources/PackageGraph/DependencyResolver.swift [hunk 1]
```diff
@@ -28,3 +29,71 @@ extension DependencyResolverError: CustomStringConvertible {
         }
     }
 }
+
+public protocol DependencyResolverDelegate {
+    func willResolve(term: Term)
+    func didResolve(term: Term, version: Version)
```
Comment (tomerd) [inline Sources/PackageGraph/DependencyResolver.swift:35]: ^^ we need for sure

File: Sources/PackageGraph/DependencyResolver.swift [hunk 2]
```diff
@@ -28,3 +29,71 @@ extension DependencyResolverError: CustomStringConvertible {
...
+
+    func derived(term: Term)
+    func conflict(conflict: Incompatibility)
+    func satisfied(term: Term, by: Assignment, incompatibility: Incompatibility)
+    func partiallySatisfied(term: Term, by: Assignment, incompatibility: Incompatibility, difference: Term)
+    func failedToResolve(incompatibility: Incompatibility)
+    func computed(bindings: [DependencyResolver.Binding])
```
Comment (tomerd) [inline Sources/PackageGraph/DependencyResolver.swift:42]: ^^ are more questionable and mirror the current logging behavior of pub grub resolver

**Comment by neonichu** [comment]
I think having a delegate that's 1:1 with the only current resolver makes sense to me, we can choose to not emit logs for every callback if we want to. I don't think we know enough to have a differently shaped delegate until we would actually have more than one resolver implementation.

**Comment by abertelrud** [comment]
> I think having a delegate that's 1:1 with the only current resolver makes sense to me, we can choose to not emit logs for every callback if we want to. I don't think we will know enough to have a differently shaped delegate until we would actually have more than one resolver implementation.

100%.  As Boris says, the callbacks should reflect what's actually happening.  Individual delegates can then easily choose to ignore some of them, as long as they are all optional.

**Review by abertelrud — APPROVED** [review]
This looks great!  I know it's a WIP but already looks very useful.

**Comment by abertelrud** [comment]
> > I think having a delegate that's 1:1 with the only current resolver makes sense to me, we can choose to not emit logs for every callback if we want to. I don't think we will know enough to have a differently shaped delegate until we would actually have more than one resolver implementation.
> 
> 100%. As Boris says, the callbacks should reflect what's actually happening. Individual delegates can then easily choose to ignore some of them, as long as they are all optional.

I think I want to modulate this statement ("all optional"):  since the delegate itself is optional, it might make sense to define the "major" events that the delegate just has to implement, and then maybe some of the minor ones could be optional (have default implementations that don't do anything).  Or should they all be optional for consistency, especially since any new ones that get added later will have to be optional for compatibility reasons (until libSwiftPM is properly versioned).  WDYT?

**Comment by tomerd** [comment]
> I think I want to modulate this statement ("all optional"): since the delegate itself is optional, it might make sense to define the "major" events that the delegate just has to implement, and then maybe some of the minor ones could be optional (have default implementations that don't do anything). Or should they all be optional for consistency, especially since any new ones that get added later will have to be optional for compatibility reasons (until libSwiftPM is properly versioned). WDYT?

I actually dislike the current pattern in SwiftPM where we have empty implementation for delegates to make them seem optional. sure, it makes life easier for the caller when they first add a delegate but it makes it very difficult to properly evolve the delegate API inc. deprecations as the default empty ones hide the fact that that API changed from the caller making these into runtime issues. in other words, I'd like to see us factor out all the default empty delegate extensions so that the caller can decide explicitly what to ignore and be notified when we deprecate delegate APIs.

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
> > I think I want to modulate this statement ("all optional"): since the delegate itself is optional, it might make sense to define the "major" events that the delegate just has to implement, and then maybe some of the minor ones could be optional (have default implementations that don't do anything). Or should they all be optional for consistency, especially since any new ones that get added later will have to be optional for compatibility reasons (until libSwiftPM is properly versioned). WDYT?
> 
> I actually dislike the current pattern in SwiftPM where we have empty implementation for delegates to make them seem optional. sure, it makes life easier for the caller when they first add a delegate but it makes it very difficult to properly evolve the delegate API inc. deprecations as the default empty ones hide the fact that that API changed from the caller making these into runtime issues. in other words, I'd like to see us factor out all the default empty delegate extensions so that the caller can decide explicitly what to ignore and be notified when we deprecate delegate APIs.

That makes sense.  That will be easier to do once we adopt semantic versioning.  At least some of the empty delegates now are to avoid breaking clients, I think.

**Comment by abertelrud** [comment]
This looks great to me.

**Review by abertelrud — APPROVED** [review]

**Pull request merged**